### PR TITLE
Do not recommend un-approved preview switches

### PR DESCRIPTION
**Pull request by Geod24**
Neither '-preview=in' nor '-previiew=revaluerefparam' has been approved by the BDFL.
'-preview=in' will require a proper DIP process before being made the default,
and DIP1016 which '-preview=revaluerefparam' implements has been rejected.
Moreover, such a message would only show up in a very narrow corner case:
in the semantic of a default argument which is 'ref' but has a rvalue.

### CODE REVIEW

**Comment by Geod24** [comment]
See conversation here: https://github.com/dlang/dmd/pull/12413#discussion_r611348007

**Review by ibuclaw — APPROVED** [review]
All tests need to be changed back though.

**Review by thewilsonator — DISMISSED** [review]
On the contrary, if people are looking for this behaviour and expect it to work (e.g. because they are coming from C++) telling them what switch they should activate to enable it is the most useful information the compiler can provide.

The exact message can be reworded to be more neutral about it but the information should still be provided.  

Also `preview=rvaluerefparam` is not exactly [unapproved](https://github.com/dlang/dmd/pull/9817)

**Comment by Geod24** [comment]
> On the contrary, if people are looking for this behaviour and expect it to work (e.g. because they are coming from C++) telling them what switch they should activate to enable it is the most useful information the compiler can provide.

See my comment, this only triggers in a very narrow case which isn't very helpful to the user, not consistent.

> The exact message can be reworded to be more neutral about it but the information should still be provided.

I disagree.

> Also `-preview=rvaluerefparam` is not exactly [unapproved](https://github.com/dlang/dmd/pull/9817)

See discussion here: https://github.com/dlang/dmd/pull/12064#issuecomment-752266308
Note that this is essentially a partial revert of #12064 . If you prefer, I can just submit a full revert of the PR.

**Comment by thewilsonator** [comment]
>  this only triggers in a very narrow case which isn't very helpful to the user, 

How so? this is much better than nothing. 

>not consistent.

That is a different problem, one that is fixed by adding more messages like this, not less.

> I disagree.

We shall have to agree to disagree, then. 

> See discussion here: #12064 (comment)

https://github.com/dlang/dmd/pull/12064#issuecomment-752271715 , https://github.com/dlang/dmd/pull/12064#issuecomment-752401478 , https://github.com/dlang/dmd/pull/12064#issuecomment-753413417 (? maybe I'm too tired atm)

I'm not sure what point you're making here.

> a partial revert of #12064 

of #12413 ? Not unless `-preview=in` is a strict superset or equal to `-preview=revaluerefparam` (see [its](https://github.com/dlang/dmd/blob/1d92ec039a43d6859005828222703d624f240d9d/src/dmd/mtype.d#L4888-L4913) [uses](https://github.com/dlang/dmd/blob/d50461f916baace1e5a9a4444e8f39651578346a/src/dmd/dtemplate.d#L1934) and [potential differences](https://github.com/dlang/dmd/pull/12242)) and `-preview=in` sets `-preview=revaluerefparam`

**Comment by Geod24** [comment]
> How so? this is much better than nothing.

It only triggers for default arguments, while in the majority of cases the users would find it more informative to see this error message on `CallExp` erroring out.

> That is a different problem, one that is fixed by adding more messages like this, not less.

Exactly, fixing the bug is a different problem from informing the user about `-preview` switch. Hence why I didn't submit a full revert. If you feel strongly about needing to inform the user about those `-preview` switch, then we should have a PR which goal is to do so, ideally consistently, instead of piggybacking on a bugfix.

> https://github.com/dlang/dmd/pull/12064#issuecomment-752271715 , https://github.com/dlang/dmd/pull/12064#issuecomment-752401478 , https://github.com/dlang/dmd/pull/12064#issuecomment-753413417 (? maybe I'm too tired atm)

Yeah, nothing about this is official. I would have no problem recommending it if there was a plan to make it the default. As it stands, the switch will be usable from this release. It obviously needs more discussion and testing before being recommended to unsuspecting user.

> of #12413 ?

Yes, that's what I meant. And the reason I mention this revert is, when in disagreement, we tend to revert. The original PR was merged while there was a disagreement on how to proceed with it. II submitted the current PR to fix this. If you don't agree with this (12425) PR, I'll just revert the original PR, because I **really** don't think we should be recommending those switches. But I hope that's not the route we're going.

**Comment by thewilsonator** [comment]
> while in the majority of cases the users would find it more informative to see this error message on CallExp erroring out.

Yes but this is `and` not, `xor`.

>  If you feel strongly about needing to inform the user about those -preview switch, then we should have a PR which goal is to do so, ideally consistently,

OK, ...

>  instead of piggybacking on a bugfix.

... (I disagree entirely with this, diagnostics should go with their fixes) ...

> because I really don't think we should be recommending those switches. 

... but [we](https://github.com/dlang/dmd/blob/master/src/dmd/parse.d#L1389) [already](https://github.com/dlang/dmd/blob/master/src/dmd/parse.d#L1405)  [do](https://github.com/dlang/dmd/blob/master/src/dmd/parse.d#L5126)

**Comment by RazvanN7** [comment]
@Geod24 @thewilsonator how should we proceed here, given #12428 ? Maybe we should continue suggesting `-preview=rvaluerefparam` , but not `-preview=in` ?

**Comment by RazvanN7** [comment]
ping @Geod24 @thewilsonator

**Comment by PetarKirov** [comment]
@RazvanN7 I think ultimate @WalterBright and @atilaneves should decide on the policy that we should follow.

**Comment by Geod24** [comment]
Although that was the point of "un-approved". Neither of those switches have seen approval from the BDFL. If it was `-preview=dip1000`, that would be a different discussion.

**Comment by 12345swordy** [comment]
@Razvan I have email Andrei regarding the dip, we are still waiting for an answer from walter for the question that asked by Andrei, regarding the rvalue dip.

**Review by WalterBright — APPROVED** [review]

**Pull request merged**
